cmd/syncthing: Decrypt versions folder (#8575)

Ratio2 opened the pull request.

tomasz1986 commented:
What if the user uses a custom versions folder? That probably should be decrypted the same way, so hard-coding `.stversions` only may not be the best idea.

acolomb commented on cmd/syncthing/decrypt/decrypt.go:
```diff
@@ -88,7 +89,7 @@ func (c *CLI) walk() error {
 			return nil
 		}
-		if fs.IsInternal(path) {
+		if fs.IsInternal(path) && !strings.HasPrefix(path, ".stversions") {
```
The string should be defined as a constant once instead of repeating it everywhere. Please check whether there is already some central place where it is defined.

It might even depend on the versioning configuration? I'm not sure whether the path can be configured for simple, trashcan and staggered versioning, but at least for external we cannot know a constant path. That should be respected in the code or at least a comment regarding the limitation.

Ratio2 replied:
No public shared space for constants, left one copy in added code

calmh replied:
We could perhaps instead look at whether the path contains `.syncthing-enc` or not

acolomb commented:
Agree with @tomasz1986, there seem to be some edge cases which require clarification. Ideally, you should use the provided PR template and describe the testing you have done, and what corner cases have been checked.

Ratio2 commented:
1. I rewrote the code to support any version folder name.
2. I did not find a public common place for constants, so I left 1 instance in my code.
3. Support for multiple versions of the same file was implemented when creating the pull request.
4. I implemented stricter path checking for data retrieval ([*/]X.syncthing-enc/XX/*[~date-time]). At the slightest inconsistency, the old behavior is used.
5. The current implementation is suitable for trash can / simple / staggered versioning. External versioning will also work if there is an intermediate path with [/]X.syncthing-enc/XX/. The date will be added if the path ends with ~date-time.

Testing:
1. I created 3 folders with the same contents (1, 1.txt, 1.txt.txt, .txt, .txt.txt)
2. I shared them to another device with different versioning (trash can / simple / staggered). Changed the version folder to versions/subver.
3. Deleted files
4. I decrypted the trash can and got:
versions/subver/1
versions/subver/1.txt
versions/subver/1.txt.txt
versions/subver/.txt
versions/subver/.txt.txt
5. I decrypted simple/staggered and got:
versions/subver/1~20221004-013833
versions/subver/1~20221004-013833.txt
versions/subver/1.txt~20221004-013833.txt
versions/subver/~20221004-013833.txt
versions/subver/.txt~20221004-013833.txt
6. The result is the same as versioning of unencrypted folders.

calmh commented on cmd/syncthing/decrypt/decrypt.go:
```diff
@@ -34,7 +36,8 @@ type CLI struct {
 
-	folderKey *[32]byte
+	folderKey             *[32]byte
+	reconstructNameRegexp *regexp.Regexp
```
This one is always the same so could be declared once at the top level

calmh commented on cmd/syncthing/decrypt/decrypt.go:
```diff
@@ -285,3 +293,27 @@ func loadEncryptedFileInfo(fd fs.File) (*protocol.FileInfo, error) {
 	return &encFi, nil
 }
+
+func (c *CLI) reconstructName(name, path string) string {
```
I'd like to see some tests on this, it's not entirely obvious to me this is correct and it would be easy to exercise in a table driven test

st-review commented:
:robot: beep boop

I'm going to close this pull request as it has been idle for more than 90 days.

This is not a rejection, merely a removal from the list of active pull requests that is periodically reviewed by humans. The pull request can be reopened when there is new activity, and merged once any remaining issues are resolved.

To permanently exempt this PR from bot nagging, add the `slow-pr` label.
